docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,75 @@
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks standardization and consistency. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a background error message or lose signal without further information.
+
+The objective with the Connection Manager is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, we have decided to encapsulate this new functionality within the Connection Manager. This object will interact with both the controller and connection health, serving as a dedicated entity for managing and improving VPN connectivity. This approach ensures a more streamlined and efficient handling of VPN connectivity issues, ultimately enhancing the user experience.
+
+## Proposed Solution
+
+We are introducing a new object, the Connection Manager, which will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionManager::State`. The Connection Manager will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful.
+
+In the event of a probe failure, the Connection Manager will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted. Once the issue has been resolved, the cycle of probes commences anew. If there are no ongoing probes, the Connection Manager will enter an Idle state.
+
+The Connection Health, an existing component of our codebase, is tasked with sending periodic pings and determining the stability of the VPN based on the responses received (see `ConnectionHealth::healthCheckup()`). It classifies the VPN stability as `Stable`, `Unstable` (some missing responses), or No Signal (no responses received). When the VPN experiences issues with network connectivity or server availability, it naturally transitions into the No Signal state and, ideally, recovers once the issue is resolved. We can leverage this logic to recover from a probe failure within the Connection Manager.
```

Comment:
> We can leverage this logic to recover from a probe failure within the Connection Manager.

Would you mind expanding a bit on how we would do that? Is ConnectionHealth listening for ConnectionManager probe failures? I see an arrow going from one to the other in the attached diagram, but the text only talk about "emmiting a signal to the front end". Is ConnectionHealth == front end in this case?

Reply:
Wait, I just re-read "We will __not__ manipulate the Connection Health stability based on Connection Manager probes". I'm confused.

Reply:
And then
>   - Wait until Connection Health stability goes back to Stable or Unstable
>  - Restart the sequence of ConnectionManager checks

So is ConnectionManager the one listening for ConnectionHealth?

Reply:
Yes correct, if a ConnectionManager probe fails and ConnectionHealth stability is NoSignal, Connection Manager aborts all probes until the stability changes to stable or unstable.

If VPN is in NoSignal we don't advance to the next probe.



For example, if the VPN is already toggled on and the internet goes down:

1. VPN goes into NoSignal via Connection Health
2. Connection Manager internet probe fails
3. Connection Manager emits a signal to the front end about network being down
4. User resolves their internet issue somehow
5. Connection Health goes back to Stable
6. Now because connection health is no longer in NoSignal, the connection manager checks restart

Reply:
The Figma dashboard really helped me understand things better. So, the connection health is like blunt tool that is just telling us whether there is signal or not. While the connection manager is a more precise tool that tells us exactly why things are weird and thus is able to provide more specific troubleshooting info to the front end. Is that correct?

Reply:
> Yes correct, if a ConnectionManager probe fails and ConnectionHealth stability is NoSignal, Connection Manager aborts all probes until the stability changes to stable or unstable.

Ok, but in the 6 steps you described there is no point in which the connection manager actually listens for signals from the connection health. Is it a case where the effects are related but not actually tied to each other?

Reply:
> While the connection manager is a more precise tool that tells us exactly why things are weird and thus is able to provide more specific troubleshooting info to the front end. Is that correct?

Yes that's right.

> in the 6 steps you described there is no point in which the connection manager actually listens for signals from the connection health. 

Step 5 and 6 are dependent on the connection health stability and where the connection manager is listening to the stability to determine when the connection manager should restart the checks

Reply:
My confusion here, IMO, stems for some poorly names classes. Specifically ConnectionManager. That name doesn't really convey "a more precise tool that tells us exactly why things are weird". I understand if this is out of context, worst case scenario take this as an FYI.

Reply:
Renamed it to `ConnectionDiagnostics` but if the team feels more excited about `ConnectionDoctor` I could be persuaded easily... 👩🏽‍⚕️

Reply:
Love `ConnectionDiagnostics`. This feels much better - the update I didn't know I was looking for, but feels perfect. Thank you.

Reply:
Thanks to Bea for expressing her dislike for the old `ConnectionManager` name which led to this one. ChatGPT had some other suggestions such as `ConnectivityDiagnosticsManager`, `NetProbe` and `NetDiagnostics` too but we settled on `ConnectionDiagnostics` 👌🏽